### PR TITLE
Use RREdgeId in t_heap

### PR DESCRIPTION
**Pull request by litghost**
#### Description

When reviewing https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1197#discussion_r389216334 @kmurray suggested removing `iconn` from the `ConnectionRouter` completely here: https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1197

I've made that change, but done it here, because removing `iconn` is more involved because of its usage in `t_heap`.

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: vpr/src/route/connection_router.cpp [hunk 1]
```diff
@@ -431,7 +431,7 @@ void ConnectionRouter::timing_driven_expand_neighbour(t_heap* current,
                        "      Pruned expansion of node %d edge %d -> %d"
                        " (to node location %d,%dx%d,%d outside of expanded"
                        " net bounding box %d,%dx%d,%d)\n",
-                       from_node, from_node_edge_idx, to_node_int,
+                       from_node, from_edge, to_node_int,
```
Comment (kmurray) [inline vpr/src/route/connection_router.cpp]: I think here, and a couple of other places we need to be calling `size_t(from_edge)` and using `%zu` in the printf pattern.

File: vpr/src/base/vpr_types.h [hunk 2]
```diff
@@ -1205,7 +1233,7 @@ struct t_trace {
  * occ:        The current occupancy of the associated rr node              */
 struct t_rr_node_route_inf {
     int prev_node;
```
Comment (kmurray) [inline vpr/src/base/vpr_types.h]: In the same vein, we should replace various references to nodes by int index with RRNodeId. Although in some cases like this, since RREdgeIds are unique we may even be able to drop storing the previous node (provided it can be looked up efficiently via the RR graph). Probably best to land this, and do the RRNodeId conversion in another PR.

File: vpr/src/route/heap_type.h [hunk 3]
```diff
@@ -43,11 +43,36 @@ struct t_heap {
 
     struct t_prev {
         int node;
-        int edge;
+        unsigned int edge;
```
Comment (kmurray) [inline vpr/src/route/heap_type.h]: Ultimately this should be replaceable with just RREdgeId (as noted earlier). Thoughts?

**Comment by kmurray** [comment]
@litghost after resolving the conflicts this should be good to merge!

File: vpr/src/route/heap_type.h (hunk 3 above)
Comment (litghost) [reply vpr/src/route/heap_type.h]: Having non-trival types in a union is hard(-ish).  It is possible, but it is really intended for things like a variant type.  I used this approach because it avoided needing to handle non-trival types under the union

File: vpr/src/base/vpr_types.h (hunk 2 above)
Comment (litghost) [reply vpr/src/base/vpr_types.h]: Ya, a follow on PR to convert all the inode to RRNodeId would be good.  But smaller PR's -> faster to review,  faster to merge is probably better.

File: vpr/src/route/connection_router.cpp (hunk 1 above)
Comment (litghost) [reply vpr/src/route/connection_router.cpp]: Yep, fixed.

**Comment by litghost** [comment]
> @litghost after resolving the conflicts this should be good to merge!

Conflicts resolved, waiting for green CI.

**Comment by litghost** [comment]
> > @litghost after resolving the conflicts this should be good to merge!
> 
> Conflicts resolved, waiting for green CI.

CI is green, merge if acceptable!
